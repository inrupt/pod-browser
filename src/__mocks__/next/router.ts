export default {
  push: jest.fn().mockResolvedValue(null),
};
