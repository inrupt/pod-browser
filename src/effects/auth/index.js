/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useEffect } from "react";
import Router from "next/router";
import { getSolidDatasetWithAcl, hasResourceAcl } from "@inrupt/solid-client";
import { useSession } from "@inrupt/solid-ui-react";
import useAuthenticatedProfile from "../../hooks/useAuthenticatedProfile";
import usePodRoot from "../../hooks/usePodRoot";

export const SESSION_STATES = {
  LOGGED_IN: "LOGGED_IN",
  LOGGED_OUT: "LOGGED_OUT",
};

export async function redirectBasedOnSessionState(
  sessionIsLoggedIn,
  sessionRequestInProgress,
  redirectIfSessionState,
  location
) {
  if (sessionRequestInProgress) {
    return;
  }

  if (
    sessionIsLoggedIn &&
    redirectIfSessionState === SESSION_STATES.LOGGED_IN
  ) {
    await Router.push(location);
  }

  if (
    !sessionIsLoggedIn &&
    redirectIfSessionState === SESSION_STATES.LOGGED_OUT
  ) {
    await Router.push(location);
  }
}

export function useRedirectIfLoggedOut(location = "/login") {
  const { session, sessionRequestInProgress } = useSession();

  useEffect(() => {
    redirectBasedOnSessionState(
      session.info.isLoggedIn,
      sessionRequestInProgress,
      SESSION_STATES.LOGGED_OUT,
      location
    );
  });
}

export function useRedirectIfLoggedIn(location = "/") {
  const { session, sessionRequestInProgress } = useSession();

  useEffect(() => {
    redirectBasedOnSessionState(
      session.info.isLoggedIn,
      sessionRequestInProgress,
      SESSION_STATES.LOGGED_IN,
      location
    );
  });
}

export function useRedirectIfNoControlAccessToOwnPod(
  resourceUrl,
  location = "/access-required"
) {
  const { session, sessionRequestInProgress } = useSession();
  const profile = useAuthenticatedProfile();
  const podRoot = usePodRoot(resourceUrl, profile);

  useEffect(() => {
    if (sessionRequestInProgress || !profile) return;

    Promise.all(
      profile.pods
        .filter((pod) => pod === podRoot)
        .map((pod) =>
          getSolidDatasetWithAcl(pod, {
            fetch: session.fetch,
          })
        )
    ).then((podResources) => {
      const hasControlAccessToPods = !podResources.find(
        (podResource) => !hasResourceAcl(podResource)
      );
      if (hasControlAccessToPods) return;
      Router.push(location);
    });
  }, [session, sessionRequestInProgress, profile, location, podRoot]);
}
