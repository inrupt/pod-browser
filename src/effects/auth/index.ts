import { useEffect } from 'react';
import Router from 'next/router';

import { getSession } from '../../../lib/solid-auth-fetcher/dist';


// TODO figure out typescript enums
export const SESSION_STATES = {
  LOGGED_IN: 'LOGGED_IN',
  LOGGED_OUT: 'LOGGED_OUT',
};

export async function redirectBasedOnSessionState(
  redirectIfSessionState: string,
  location: string,
  getSAFSession: Function,
) {
  const session = await getSAFSession();

  if (!session && redirectIfSessionState === SESSION_STATES.LOGGED_OUT) {
    Router.push(location);
  }

  if (session && redirectIfSessionState === SESSION_STATES.LOGGED_IN) {
    Router.push(location);
  }
}

export function useRedirectIfLoggedOut(
  location = '/login',
  getSAFSession = getSession,
) {
  return useEffect(() => {
    redirectBasedOnSessionState(SESSION_STATES.LOGGED_OUT, location, getSAFSession);
  }, [location, getSAFSession]);
}

export function useRedirectIfLoggedIn(
  location = '/',
  getSAFSession = getSession,
) {
  return useEffect(() => {
    redirectBasedOnSessionState(SESSION_STATES.LOGGED_IN, location, getSAFSession);
  }, [location, getSAFSession]);
}
