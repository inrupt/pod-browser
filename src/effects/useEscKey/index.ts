import { useEffect } from "react";

export const ESC_KEYCODE = 27;

export function handleEscKeydown(
  callback: (evnt: KeyboardEvent) => unknown
): (evnt: KeyboardEvent) => unknown {
  return (e: KeyboardEvent) => {
    if (!e?.keyCode) return;
    if (e.keyCode === ESC_KEYCODE) callback(e);
  };
}

export default function useEscKey(
  callback: (evnt: KeyboardEvent) => unknown
): void {
  return useEffect(() => {
    document.body.addEventListener("keydown", handleEscKeydown(callback));

    return () => {
      document.body.removeEventListener("keydown", handleEscKeydown(callback));
    };
  });
}
