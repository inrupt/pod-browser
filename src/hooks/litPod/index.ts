/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable camelcase, @typescript-eslint/no-explicit-any */

import useSWR from "swr";
import {
  fetchLitDataset,
  getThingOne,
  getIriAll,
  unstable_Access,
  unstable_fetchResourceInfoWithAcl,
  unstable_getAgentAccessAll,
  isContainer, Iri, stringAsIri,
} from "@solid/lit-pod";
import { LDP } from "@solid/lit-vocab-common";
import { WS } from "@solid/lit-vocab-solid";
import {
  fetchResourceWithAcl,
  getIriPath,
  normalizePermissions,
  ResourceDetails,
} from "../../lit-solid-helpers";

export async function fetchContainerResourceIris(
  containerIri: Iri
): Promise<Iri[]> {
  const litDataset = await fetchLitDataset(containerIri);
  const container = getThingOne(litDataset, containerIri);
  const iris = getIriAll(container, LDP.contains);
  return iris;
}

export const GET_CONTAINER_RESOURCE_IRIS = stringAsIri("getContainerResourceIris");
export function useFetchContainerResourceIris(iri: Iri): any {
  return useSWR<Iri[]>(
    [iri, GET_CONTAINER_RESOURCE_IRIS],
    fetchContainerResourceIris
  );
}

export async function fetchResourceDetails(
  iri: Iri
): Promise<ResourceDetails> {
  const name = getIriPath(iri) as string;
  const resourceInfo = await unstable_fetchResourceInfoWithAcl(iri);
  const accessModeList = unstable_getAgentAccessAll(resourceInfo);
  const permissions = await normalizePermissions(
    accessModeList as Record<string, unstable_Access>
  );

  let types = [] as string[];
  const contentType = resourceInfo?.resourceInfo?.contentType;

  if (contentType) types = [contentType];
  if (isContainer(resourceInfo)) types = ["Container"];

  return {
    iri,
    permissions,
    types,
    name,
  };
}

export const FETCH_RESOURCE_DETAILS = stringAsIri("fetchResourceDetails");
export function useFetchResourceDetails(iri: Iri): any {
  return useSWR([iri, FETCH_RESOURCE_DETAILS], fetchResourceDetails);
}

export const FETCH_RESOURCE_WITH_ACL = stringAsIri("fetchResourceWithAcl");
export function useFetchResourceWithAcl(iri: Iri): any {
  return useSWR([iri, stringAsIri("fetchResourceWithAcl")], fetchResourceWithAcl);
}

export async function fetchPodIrisFromWebId(webId: Iri): Promise<Iri[]> {
  const profileDoc = await fetchLitDataset(webId);
  const profile = getThingOne(profileDoc, webId);

  return getIriAll(profile, WS.storage);
}
export const FETCH_POD_IRIS_FROM_WEB_ID = stringAsIri("fetchPodIrisFromWebId");
export function useFetchPodIrisFromWebId(webId: Iri): any {
  return useSWR([webId, FETCH_POD_IRIS_FROM_WEB_ID], fetchPodIrisFromWebId);
}
