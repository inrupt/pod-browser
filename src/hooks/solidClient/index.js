/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable camelcase */
import { useContext } from "react";
import useSWR from "swr";
import { space, ldp } from "rdf-namespaces";

import {
  fetchLitDataset,
  fetchResourceInfoWithAcl,
  getIriAll,
  getResourceAcl,
  getThing,
  hasResourceAcl,
  isContainer,
  unstable_getAgentAccessAll,
  unstable_getAgentDefaultAccessAll,
} from "@inrupt/solid-client";

import { getIriPath } from "../../solidClientHelpers/utils";
import { normalizePermissions } from "../../solidClientHelpers/permissions";
import { fetchResourceWithAcl } from "../../solidClientHelpers/resource";

import SessionContext from "../../contexts/sessionContext";

export async function fetchContainerResourceIris(containerIri, fetch) {
  const litDataset = await fetchLitDataset(containerIri, { fetch });
  const container = getThing(litDataset, containerIri);
  const iris = getIriAll(container, ldp.contains);
  return iris;
}

export const GET_CONTAINER_RESOURCE_IRIS = "getContainerResourceIris";
/* istanbul ignore next */
export function useFetchContainerResourceIris(iri) {
  const { session } = useContext(SessionContext);

  return useSWR([iri, GET_CONTAINER_RESOURCE_IRIS], () =>
    fetchContainerResourceIris(iri, session.fetch)
  );
}

export async function fetchResourceDetails(iri, fetch) {
  const name = getIriPath(iri);
  const resourceInfo = await fetchResourceInfoWithAcl(iri, { fetch });
  let defaultPermissions = [];
  let permissions = [];

  if (hasResourceAcl(resourceInfo)) {
    const accessModeList = unstable_getAgentAccessAll(resourceInfo);
    permissions = await normalizePermissions(accessModeList, fetch);
    const resourceAcl = getResourceAcl(resourceInfo);
    const defaultAccessModeList = unstable_getAgentDefaultAccessAll(
      resourceAcl
    );

    defaultPermissions = await normalizePermissions(
      defaultAccessModeList,
      fetch
    );
  }

  let types = [];
  const contentType = resourceInfo?.internal_resourceInfo?.contentType;

  if (contentType) types = [contentType];
  if (isContainer(resourceInfo)) types = ["Container"];

  return {
    iri,
    permissions,
    defaultPermissions,
    types,
    name,
    dataset: resourceInfo,
  };
}

export const FETCH_RESOURCE_DETAILS = "fetchResourceDetails";
/* istanbul ignore next */
export function useFetchResourceDetails(iri) {
  const { session } = useContext(SessionContext);
  return useSWR([iri, FETCH_RESOURCE_DETAILS], () =>
    fetchResourceDetails(iri, session.fetch)
  );
}

export const FETCH_RESOURCE_WITH_ACL = "fetchResourceWithAcl";
/* istanbul ignore next */
export function useFetchResourceWithAcl(iri) {
  const { session } = useContext(SessionContext);
  return useSWR([iri, FETCH_RESOURCE_WITH_ACL], () =>
    fetchResourceWithAcl(iri, session.fetch)
  );
}

export async function fetchPodIrisFromWebId(webId, fetch) {
  const profileDoc = await fetchLitDataset(webId, { fetch });
  const profile = getThing(profileDoc, webId);

  return getIriAll(profile, space.storage);
}

export const FETCH_POD_IRIS_FROM_WEB_ID = "fetchPodIrisFromWebId";
/* istanbul ignore next */
export function useFetchPodIrisFromWebId(webId) {
  const { session } = useContext(SessionContext);
  return useSWR([webId, FETCH_POD_IRIS_FROM_WEB_ID], () =>
    fetchPodIrisFromWebId(webId, session.fetch)
  );
}
