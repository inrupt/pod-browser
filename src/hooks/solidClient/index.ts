/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable camelcase, @typescript-eslint/no-explicit-any */

import useSWR from "swr";
import {
  fetchLitDataset,
  getThingOne,
  getIriAll,
  unstable_Access,
  unstable_fetchResourceInfoWithAcl,
  unstable_getAgentAccessAll,
  isContainer,
} from "@inrupt/solid-client";
import { space } from "rdf-namespaces";
import {
  fetchResourceWithAcl,
  getIriPath,
  namespace,
  normalizePermissions,
  IResourceDetails,
} from "../../solidClientHelpers";

export async function fetchContainerResourceIris(
  containerIri: string
): Promise<string[]> {
  const litDataset = await fetchLitDataset(containerIri);
  const container = getThingOne(litDataset, containerIri);
  const iris = getIriAll(container, namespace.contains);
  return iris;
}

export const GET_CONTAINER_RESOURCE_IRIS = "getContainerResourceIris";
export function useFetchContainerResourceIris(iri: string): any {
  return useSWR<string[]>(
    [iri, GET_CONTAINER_RESOURCE_IRIS],
    fetchContainerResourceIris
  );
}

export async function fetchResourceDetails(
  iri: string
): Promise<IResourceDetails> {
  const name = getIriPath(iri) as string;
  const resourceInfo = await unstable_fetchResourceInfoWithAcl(iri);
  const accessModeList = unstable_getAgentAccessAll(resourceInfo);
  const permissions = await normalizePermissions(
    accessModeList as Record<string, unstable_Access>
  );

  let types = [] as string[];
  const contentType = resourceInfo?.resourceInfo?.contentType;

  if (contentType) types = [contentType];
  if (isContainer(resourceInfo)) types = ["Container"];

  return {
    iri,
    permissions,
    types,
    name,
  };
}

export const FETCH_RESOURCE_DETAILS = "fetchResourceDetails";
export function useFetchResourceDetails(iri: string): any {
  return useSWR([iri, FETCH_RESOURCE_DETAILS], fetchResourceDetails);
}

export const FETCH_RESOURCE_WITH_ACL = "fetchResourceWithAcl";
export function useFetchResourceWithAcl(iri: string): any {
  return useSWR([iri, "fetchResourceWithAcl"], fetchResourceWithAcl);
}

export async function fetchPodIrisFromWebId(webId: string): Promise<string[]> {
  const profileDoc = await fetchLitDataset(webId);
  const profile = getThingOne(profileDoc, webId);

  return getIriAll(profile, space.storage);
}
export const FETCH_POD_IRIS_FROM_WEB_ID = "fetchPodIrisFromWebId";
export function useFetchPodIrisFromWebId(webId: string): any {
  return useSWR([webId, FETCH_POD_IRIS_FROM_WEB_ID], fetchPodIrisFromWebId);
}
