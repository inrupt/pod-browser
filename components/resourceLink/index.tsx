/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { ReactNode, ReactElement, useContext } from "react";
import { NextRouter } from "next/router";
import Link from "next/link";
import PodLocationContext from "../../src/contexts/podLocationContext";

export function resourceHref(iri: string): string {
  return `/resource/${encodeURIComponent(iri)}`;
}

// TODO make typescript happy someday

export const urlForResourceAction = (
  action: string | undefined,
  resourceIri: string,
  containerIri: string | undefined
): any => ({
  pathname: containerIri ? resourceHref(containerIri) : "/resource/[iri]",
  ...(action
    ? {
        query: {
          action,
          resourceIri,
        },
      }
    : {}),
});

export function resourceContextRedirect(
  action: string,
  resourceIri: string,
  containerIri: string,
  router: NextRouter
): Promise<boolean> {
  return router.replace(
    urlForResourceAction(action, resourceIri, undefined),
    urlForResourceAction(action, resourceIri, containerIri)
  );
}

/* eslint react/require-default-props: 0 */
interface IResourceLink {
  children?: ReactNode;
  action?: string;
  resourceIri?: string;
  containerIri?: string;
  ref?: any; // TODO no idea
}

/* eslint react/jsx-props-no-spreading: 0 */
export default function ResourceLink(
  props: IResourceLink
): ReactElement | null {
  const { children, action, resourceIri, containerIri, ...linkProps } = props;

  // If resource and container aren't passed in, create a link to show <action> for the current uri.
  const { currentUri } = useContext(PodLocationContext);
  const resourceIriToDisplay = resourceIri || currentUri;
  const containerIriToUseAsBase = containerIri || currentUri;

  return (
    <Link
      href={urlForResourceAction(action, resourceIriToDisplay, undefined)}
      as={urlForResourceAction(
        action,
        resourceIriToDisplay,
        containerIriToUseAsBase
      )}
      replace
    >
      <a {...linkProps}>{children}</a>
    </Link>
  );
}
