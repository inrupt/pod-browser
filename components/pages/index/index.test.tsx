/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import * as nextRouterFns from "next/router";

import { useRedirectIfLoggedOut } from "../../../src/effects/auth";
import { useFetchPodIrisFromWebId } from "../../../src/hooks/solidClient";
import { resourceHref } from "../../resourceLink";
import IndexPage from "./index";

jest.mock("../../../src/effects/auth");
jest.mock("../../../src/hooks/solidClient");
jest.mock("@inrupt/solid-client");
jest.mock("next/router");

describe("Index page", () => {
  test("Renders null if there are no pod iris", () => {
    (nextRouterFns.useRouter as jest.Mock).mockReturnValue({
      replace: jest.fn().mockResolvedValue(undefined),
    });

    (useFetchPodIrisFromWebId as jest.Mock).mockReturnValue({
      data: undefined,
    });

    const tree = shallow(<IndexPage />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  test("Redirects to the resource page if there is a pod iri", () => {
    const replace = jest.fn().mockResolvedValue(undefined);
    const podIri = "https://mypod.myhost.com";

    (nextRouterFns.useRouter as jest.Mock).mockReturnValue({ replace });

    (useFetchPodIrisFromWebId as jest.Mock).mockReturnValue({
      data: [podIri],
    });

    shallow(<IndexPage />);

    expect(replace).toHaveBeenCalledWith(
      "/resource/[iri]",
      resourceHref(podIri)
    );
  });

  test("Redirects if the user is logged out", () => {
    shallow(<IndexPage />);
    expect(useRedirectIfLoggedOut).toHaveBeenCalled();
  });
});
