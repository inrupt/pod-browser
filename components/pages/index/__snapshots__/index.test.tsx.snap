// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page Renders a logout button 1`] = `
<WithStyles(ForwardRef(Container))>
  <Home />
</WithStyles(ForwardRef(Container))>
`;
